### PR TITLE
feat: Metrics/1803 emit error viewed event

### PR DESCRIPTION
**Pull request by Cal-L**
## **Description**

This PR sends a Segment event when a user encounters an error boundary
Event name: `Error Screen Viewed`
Properties:
- `error: string` - Error message
- `boundary: string` - The boundary that a user encountered
- `stack: string` - Concatenated string with stacks 5 levels deep

## **Related issues**

Fixes: [#1803](https://github.com/orgs/MetaMask/projects/60/views/6?pane=issue&itemId=63761525)

## **Manual testing steps**

1. Open the app
2. Enable metametrics
3. Hardcode a thrown error to trigger the error boundary
4. Check `MetaMask Mobile [Dev]` source in Segment and search for `Error Screen Viewed`
5. Event should exist along with properties

### **After**

image

image

image

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by Cal-L** [comment]
@gauthierpetetin There wasn't any properties specified in the original ticket. I figured that it would be useful to include relevant information for the error boundary in the properties. Please checkout the description and LMK if you want to keep those.

**Comment by gauthierpetetin** [comment]
Thanks @Cal-L , what you suggest sounds good to me.
cc @Gudahtt in case he has additional suggestions

**Comment by NicolasMassart** [comment]
@Cal-L could you check how readable the stacktrace is?
I wonder if it allows to easily point to actual RN code or if it's bundled or native code in it, that would require some mapping.

**Review by NicolasMassart — COMMENTED** [review]
@Cal-L could you check how readable the stacktrace is?
I wonder if it allows to easily point to actual RN code or if it's bundled or native code in it, that would require some mapping.

EDIT: discussed with @Cal-L and the stack is just a string so should be readable and is trimmed so no risk to overflow the Segment events limit.

File: app/components/Views/ErrorBoundary/index.js [hunk 1]
```diff
@@ -184,13 +188,31 @@ class ErrorBoundary extends Component {
...
   };
 
   static getDerivedStateFromError(error) {
     return { error };
   }
 
+  generateErrorReport = (error, errorInfo = '') => {
+    const {
+      view,
+      metrics: { trackEvent },
```
Comment (NicolasMassart) [inline app/components/Views/ErrorBoundary/index.js:201]: C: I did not know we could to this, that's simplifying the code, thanks for sharing!

**Review by NicolasMassart — DISMISSED** [review]
Looks good to me.

**Review by NicolasMassart — DISMISSED** [review]
Looks good to me.

File: app/components/Views/ErrorBoundary/index.test.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  }, []);
+  return <View />;
+};
+
+describe('ErrorBoundary', () => {
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it('should render correctly', () => {
```
Comment (NicolasMassart) [inline app/components/Views/ErrorBoundary/index.test.tsx:29]:
```suggestion
  it('render matches snapshot', () => {
```

**Review by NicolasMassart — DISMISSED** [review]
Looks good to me.

**Review by NicolasMassart — APPROVED** [review]

**Pull request merged**